src/Workspaces/Core/Portable/Workspace/Workspace_EventsLegacy.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis;
+
+public abstract partial class Workspace
+{
+    // Allows conversion of legacy event handlers to the new event system. The first item in
+    // the key's tuple is an EventHandler<TEventArgs> and thus stored as an object.
+    private readonly Dictionary<(object EventHandler, WorkspaceEventType EventType), (int AdviseCount, IDisposable Disposer)> _disposableEventHandlers = new();
+    private readonly object _gate = new();
```

Comment:
Given this Workspace type already has a lot of partials, consider making this _gate a name more like _legacyWorkspaceEventsGate before somebody else "helpfully" acquires this lock for an unrelated reason.